### DISCUSSION TITLE
Rebasing/merging intelligently during scheduled time

### DISCUSSION THREAD

**Discussion by mayfieldiv** [renovatebot/renovate#23007] [Suggest an Idea]
First off, thank you for making and supporting this great product!

**What would you like Renovate to be able to do?**

The current behavior of the following config during the scheduled time can be a bit inefficient if there are multiple branches to be merged. As soon as one branch is merged, all the others are rebased. As soon as one of rebased branches passes it's status checks, it's merged and the other branches are rebased.

```json
{
  "extends": ["config:base", ":automergeMinor", "schedule:nonOfficeHours"],
  "rebaseStalePrs": true,
  "prCreation": "immediate",
  "stabilityDays": 3
}
```

It'd be nice if Renovate didn't rebase all eligible branches at the same time, since after every branch it merges, it has to rebase every other branch again, potentially using a lot of extra time/resources due to the CI pipeline being clogged up

**Describe the solution you'd like**

Given the above config, I think the ideal behavior would as described in the [comment here](https://github.com/renovatebot/renovate/issues/4806#issuecomment-553072900):

**Describe alternatives you've considered**

Based on the documentation, it's possible setting `rebaseStalePrs` to `false` and `automergeType` to `branch` might achieve better behavior when it comes time to merge branches, but that's not a great solution because I still want to see the PR for a few days so I can know what changes are upcoming.

I tried setting `rebaseStalePrs` to either `null` or `false` but that caused Renovate to not rebase when needed to merge when `automergeType` was `pr`. I have my GitLab repo set to only allow fast-forward merges, so a branch about to merge must be at the head of master.

**Comment by rarkins** [comment]
Hi @mayfieldiv, I've move this to the main Renovate repo as I think your idea for custom rebasing (only one at a time, etc) will need feature addition rather than merely config. Marking it also as `needs requirements` as I think we'll need to circle around a few times to think of all the possibilities and edge cases.

For example:

> During the scheduled time, renovate will go to the first merge request where the status checks all pass (including stability), and merge it into master, rebasing (recreating) as necessary

"rebasing (recreating) as necessary" is the key phrase here, because every time we do that, it triggers the tests to run again. So we can't do literally as you describe in the case where the branch has fallen behind `master`.

**Comment by mayfieldiv** [comment]
An example may help clarify:

Say we've just entered the schedule window. The current state of the branches are as follows:
Branch A: meets stability requirements, tests pass, **behind** master
Branch B: does **NOT** meets stability requirements, tests pass, **behind** master
Branch C: does **NOT** meets stability requirements, tests pass, **behind** master
Branch D: meets stability requirements, tests pass, **behind** master
Branch E: meets stability requirements, tests pass, on head of master

The way I'm thinking it might work is:
1. Branch E is merged
2. Branch A is rebased
3. Wait until Branch A pipeline is complete
4. If pipeline passes, merge Branch A. Otherwise, move on
5. Same three steps (2, 3, 4) occur for Branch D
6. Branch B is rebased, not waiting for pipeline
7. Branch C is rebased, not waiting for pipeline

I've skipped other steps regarding checking dependencies for updates, updating the master issue, etc. But the idea is that each time renovate runs, it checks the state of the branches and prioritizes in the above manner.

**Comment by rarkins** [comment]
This is not an impossible dream but quite a change from the current approach. It looks like it would require multiple "passes" over the branches on each run to determine the best actions(s) to take for that run.

Also probably each of those steps needs to be done in a separate "run" of Renovate, with the need for at each step to clearly define "at this point in time, how would Renovate know to do X?". Something like:
- Automerge the *first* branch that is up to date and has passed tests (note: I'd consider stability to be part of tests). Abort rest of run after automerging.
- If any branches are up to date but still running tests, do nothing
- If no branches are ready for automerge, rebase the first that passes tests

Step 3 in yours is perhaps the hardest one - knowing that we're waiting on A to pass tests and skipping it. If we keep the stability check as a "test" then it would mean we do nothing for example 3 days waiting for it to reach stability. So maybe we need to treat stability as a special case and not a regular test.

**Comment by mayfieldiv** [comment]
Yeah I get what you're saying. I'm happy to continue discussing how this feature request would work, such as how configuration options could augment its behavior, but also what do you think about the other parts of my initial description of the config I'm trying to reach?

For example, do you think I should be able to set `"rebaseStalePrs": null` and Renovate rebases once all status checks pass if the repo is set to require fast-forward merges?

I guess a lot of my requests center around a different behavior of the `schedule` option. Where the schedule window is the time to rebase and merge, but everything else still happens outside the schedule. I see there exists `updateNotScheduled` to augment the behavior of `schedule`, but maybe a more granular way to specify what actions to include/exclude from non-scheduled times could be good (maybe another feature request).

**Comment by mayfieldiv** [comment]
Is there anything I can do to help organize this ticket or otherwise further the conversation? Like open another ticket to narrow the scope of this one?

**Comment by rarkins** [comment]
@mayfieldiv actually yes, for the other config questions could you post them separately back in the config-help repo? It's easiest to address them separately

**Comment by geraintwhite** [comment]
This sounds like a great idea and something I would like to see added.

Another config option I would like to see is the ability to rebase other branches after renovate is finished.

e.g.

- Merge all passing passing renovate branches are into the base branch.
- Go through each of the other open non-renovate PRs with the same base branch and rebase them onto the base branch.

This would reduce the CI overhead of rebasing all PRs during office hours.

**Comment by rarkins** [comment]
> * Merge all passing passing renovate branches are into the base branch.

Not as easy as it sounds. Setting aside the risk that merge of an update to package A into master could actually break the update to B, that update to A might also result in a conflict in B's branch. Even if not, platforms like GitHub can take seconds to recalculate, so temporarily return "not ready for merge" results for B and other branches if Renovate tried them too quickly. You'd need some arbitrary delay like 10 seconds after automerging and even then could not be sure that you wouldn't get an error.

> * Go through each of the other open non-renovate PRs with the same base branch and rebase them onto the base branch.

You mean you want Renovate to attempt to rebase essentially all PRs once it's automerged, not just its own?

**Comment by geraintwhite** [comment]
For the first point, I mean wait until all renovate branches are automerged through the normal process. Then after that has happened, start rebasing non-renovate branches.

**Comment by mayfieldiv** [comment]
It seems like doing anything to non-Renovate branches should be out of scope for a dependency updater such as Renovate

**Comment by rarkins** [comment]
#10229 made a small improvement here where it will now immediately re-run on the repository a second time after any automerge, which should mean other branches get rebased within seconds.
